widget/diagramwidget/diagram.go

```diff
@@ -0,0 +1,251 @@
+package diagramwidget
+
+import (
+	"image/color"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+	"fyne.io/fyne/v2/driver/desktop"
+	"fyne.io/fyne/v2/theme"
+	"fyne.io/fyne/v2/widget"
+)
+
+var Globaldiagram *DiagramWidget
+
+// ForceRepaint is a workaround for a Fyne bug (Issue #2205) in which moving a canvas object does not
+// trigger repainting. When the issue is resolved, this function and all references to it should be
+// removed. The DummyBox on the GlobalDiagram should also be removed.
+// The conditionals here are required during initialization.
+func ForceRepaint() {
+	if Globaldiagram != nil {
+		if Globaldiagram.DummyBox != nil {
+			Globaldiagram.DummyBox.Refresh()
+		}
+	}
+}
+
+// Verify that interfaces are fully implemented
+var _ fyne.Tappable = (*DiagramWidget)(nil)
+
+type linkPinPair struct {
+	link *DiagramLink
+	pin  *LinkPoint
+}
+
+type DiagramWidget struct {
+	widget.BaseWidget
+
+	// ID is expected to be unique across all DiagramWidgets in the application.
+	ID string
```

Comment:
Is this needed? the `DiagramWidget` itself is unique, an additional ID should not be needed.

Reply:
ID is needed here in the same way that ID is needed for tree widget entries: to facilitate a) retrieving widgets and b) mapping widgets to the underlying information they represent (which is often more complicated than a simple value
binding).